### PR TITLE
start separating js.t from js_internal.t

### LINKED ISSUE #782: Js__.Js_internal leaks into type errors and LSP diagnostics
Since Melange v2 with stdlib being wrapped on each package (https://github.com/melange-re/melange/pull/624) there's a few places where errors got worst.

```reason
let obj = Js.Obj.empty();
let _ = Shared_js.MelRaw.send(obj);
```

```
Line 4, 13:
Error This expression has type < .. > Js__.Js_internal.t
but an expression was expected of type string
``` 

## v2

https://melange.re/unstable/playground/?language=Reason&code=bGV0IHNlbmQgPSAob2JqOiBzdHJpbmcpID0%2BIFslbWVsLnJhdyAiZnVuY3Rpb24gKG9iaikgeyBjb25zb2xlLmxvZyhvYmopIH0iXTsKCmxldCBvYmogPSBKcy5PYmouZW1wdHkoKTsKbGV0IF8gPSBzZW5kKG9iaik7Cg%3D%3D&live=on

## v1

https://melange.re/v1.0.0/playground/?language=Reason&code=bGV0IHNlbmQgPSAob2JqOiBzdHJpbmcpID0%2BIFslYnMucmF3ICJmdW5jdGlvbiAob2JqKSB7IGNvbnNvbGUubG9nKG9iaikgfSJdOwoKbGV0IG9iaiA9IEpzLk9iai5lbXB0eSgpOwpsZXQgXyA9IHNlbmQob2JqKTsK&live=on

### PR DESCRIPTION
**Pull request by jchavarri**
Fixes #782.

### CODE REVIEW

File: ppx/ast_literal.ml
Comment (jchavarri) [inline ppx/ast_literal.ml:33]: I guess part of this todo is being done here, not sure if it can be removed.

File: jscomp/runtime/js.mli
Comment (jchavarri) [inline jscomp/runtime/js.mli:86]: This is prob the most controversial part of the PR, but before it was already accessible to consumers through `Js__.Js_OO.unsafe_downgrade`, so nothing has changed in terms of visibility.

Comment (anmonteiro) [reply jscomp/runtime/js.mli]: Ah, and we need it in `js.ml` because it needs `Js.t`. That's fine.

File: jscomp/runtime/js.ml
Comment (anmonteiro) [inline jscomp/runtime/js.ml:67]: continuing conversation from https://github.com/melange-re/melange/pull/798#discussion_r1366950903 because I deleted the `.mli` file.

I think we should expose this in a private module and explain why. I'll add a commit and we can see if you agree.

Comment (anmonteiro) [reply jscomp/runtime/js.ml]: What do you think about ba8352954 ?

**Review by anmonteiro — APPROVED** [review]
This looks good if you like my changes.

Thanks for figuring this out! This an impactful user-facing change. I also think it deserves a changelog entry.

**Comment by jchavarri** [comment]
> continuing conversation from https://github.com/melange-re/melange/pull/798#discussion_r1366950903 because I deleted the .mli file.

I am not sure I understand why it's preferable to have the `.mli` file removed. I thought it would be beneficial to include an interface in a such a fundamental module like `Js`.

**Comment by anmonteiro** [comment]
> I am not sure I understand why it's preferable to have the `.mli` file removed. I thought it would be beneficial to include an interface in a such a fundamental module like `Js`.

In this case, because it wasn't adding much. It was just repeating the externals and adding module aliases. There were no signatures apart from Js_obj

File: ppx/ast_literal.ml
Comment (anmonteiro) [reply ppx/ast_literal.ml:33]: I suppose we can remove it.

**Review by jchavarri — COMMENTED** [review]
sorry, i had 3 inline comments pending but i forgot to submit the review :/

File: jscomp/runtime/js.ml
Comment (anmonteiro) [reply jscomp/runtime/js.ml:176]: so Js_OO is not exposed at all. this allows it to be exposed in `Js.Private.Js_OO.*` and have the `Ast_literal` module be the same for every usage of `Js_OO`.

File: jscomp/runtime/js.ml
Comment (anmonteiro) [reply jscomp/runtime/js.ml:174]: it _is_ private, because it's not exposed as part of the public API. nothing ever uses `Js_OO` other than the compiler internals, right?

File: jscomp/runtime/js.ml
Comment (anmonteiro) [reply jscomp/runtime/js.ml:175]: see the answer in https://github.com/melange-re/melange/pull/798#discussion_r1367807067 and https://github.com/melange-re/melange/pull/798#discussion_r1367807213

**Comment by anmonteiro** [comment]
Whatever we do regarding the `.mli` needs to play well with the generated odoc documentation.

@jchavarri I'll leave that to you but feel free to request my help if you prefer

File: jscomp/runtime/js.ml
Comment (jchavarri) [reply jscomp/runtime/js.ml:174]: I must be missing something. I mean it is public in the sense that any user of the library can access it:

```ocaml
module T = Js.Private.Js_OO
let t = Js.Private.unsafe_downgrade
```

This is not what "private" means to me :) but I guess it's not that important.

**Comment by jchavarri** [comment]
> Whatever we do regarding the `.mli` needs to play well with the generated odoc documentation.

I agree. I think it'll be better to tackle that when we focus on improving the odoc comments more broadly.

File: jscomp/runtime/js.ml
Comment (anmonteiro) [reply jscomp/runtime/js.ml:174]: it's private as in it's internal code to the library that it's not part of the _public_ API. I'm using a pattern that I've seen in other code, e.g. https://github.com/search?q=org%3Ajanestreet%20%22module%20private%22&type=code

**Comment by anmonteiro** [comment]
Don't forget the changes entry.

File: jscomp/runtime/js.ml
Comment (jchavarri) [reply jscomp/runtime/js.ml:174]: I see, so the name is a way to signal intent to the outside world ("use this at your own risk" etc).

**Comment by jchavarri** [comment]
Couple of thoughts before we merge. Would anything be simplified if we moved the `Js_OO` module and the downgrade function to the ppx itself as a runtime library? I don't think it's used anywhere inside the `melange.js` library.

Also, will the change in this PR break anything in melange-compiler-libs? It seems to expect a specific path for `Js_OO` (without the `Private` namespacing):

https://github.com/melange-re/melange-compiler-libs/blob/eb8207c1b6744180095184194dfd843e06f9fad1/typing/oprint.ml#L309-L321

**Comment by anmonteiro** [comment]
> Would anything be simplified if we moved the Js_OO module and the downgrade function to the ppx itself as a runtime library? I don't think it's used anywhere inside the melange.js library.

well, _this_ is the runtime library. I think it's in the right place.



> Also, will the change in this PR break anything in melange-compiler-libs? It seems to expect a specific path for `Js_OO` (without the `Private` namespacing):

well... that's very well spotted, but that's already broken since the module is in fact `Js__Js_OO`, so we would need to fix it regardless.

**Comment by jchavarri** [comment]
I created https://github.com/melange-re/melange/issues/806 to track the 2nd point.

**Pull request merged**
